### ISSUE TITLE
Closing connection hangs in dpiConn_release

### ISSUE THREAD

**Issue by mguggi** [oracle/odpi#50]
I use the go-goracle/goracle driver for a golang project. This version uses the version 2.0.3 of ODPI-C. The program is running on  Centos 7 and I'm use the Instant Client 12.2.0.1 from Oracle. 

The program is running as expected, however when I try to stop the program with pressing `Ctrl+C` the program is calling my `Close()` method as expected but than the program hangs forever.

After a while (approximately 10 minutes) I open a second terminal and send `kill -ABRT <pid>` and than the stacktrace below will be printed.

```
goroutine 62 [syscall, locked to thread]:
github.com/mguggi/oraclebeat/module/oracle/vendor/gopkg.in/goracle%2ev2._Cfunc_dpiConn_release(0x46da9c0, 0x0)
	github.com/mguggi/oraclebeat/module/oracle/vendor/gopkg.in/goracle.v2/_obj/_cgo_gotypes.go:1448 +0x49
github.com/mguggi/oraclebeat/module/oracle/vendor/gopkg.in/goracle%2ev2.(*conn).Close.func1(0x46da9c0, 0x2f5d950)
	/home/mguggi/work/beats/src/github.com/mguggi/oraclebeat/module/oracle/vendor/gopkg.in/goracle.v2/conn.go:112 +0x60
github.com/mguggi/oraclebeat/module/oracle/vendor/gopkg.in/goracle%2ev2.(*conn).Close(0xc420230300, 0x0, 0x0)
	/home/mguggi/work/beats/src/github.com/mguggi/oraclebeat/module/oracle/vendor/gopkg.in/goracle.v2/conn.go:112 +0xa9
database/sql.(*driverConn).finalClose.func2()
	/home/mguggi/public/go-1.9.2/src/database/sql/sql.go:467 +0x49
database/sql.withLock(0x405f9e0, 0xc420011ce0, 0xc42019bd38)
	/home/mguggi/public/go-1.9.2/src/database/sql/sql.go:2867 +0x65
database/sql.(*driverConn).finalClose(0xc420011ce0, 0xc42019be58, 0xc42019be48)
	/home/mguggi/public/go-1.9.2/src/database/sql/sql.go:465 +0x128
database/sql.(finalCloser).(database/sql.finalClose)-fm(0xc42008d108, 0xc420346ea0)
	/home/mguggi/public/go-1.9.2/src/database/sql/sql.go:564 +0x2f
database/sql.(*DB).Close(0xc42008d0e0, 0x0, 0x0)
	/home/mguggi/public/go-1.9.2/src/database/sql/sql.go:678 +0x31b
github.com/mguggi/oraclebeat/module/oracle/sgastats.(*MetricSet).Close(0xc42008d7c0, 0x406ec20, 0xc42008d7c0)
	/home/mguggi/go/src/github.com/mguggi/oraclebeat/module/oracle/sgastats/sgastats.go:110 +0xd7
github.com/mguggi/oraclebeat/vendor/github.com/elastic/beats/metricbeat/mb/module.(*metricSetWrapper).close(0xc420118d40, 0x0, 0x0)
	/home/mguggi/go/src/github.com/mguggi/oraclebeat/vendor/github.com/elastic/beats/metricbeat/mb/module/wrapper.go:257 +0x64
github.com/mguggi/oraclebeat/vendor/github.com/elastic/beats/metricbeat/mb/module.(*Wrapper).Start.func1(0xc4200193e0, 0xc420020ba0, 0xc4202504e0, 0xc420118d40)
	/home/mguggi/go/src/github.com/mguggi/oraclebeat/vendor/github.com/elastic/beats/metricbeat/mb/module/wrapper.go:111 +0xbe
created by github.com/mguggi/oraclebeat/vendor/github.com/elastic/beats/metricbeat/mb/module.(*Wrapper).Start
	/home/mguggi/go/src/github.com/mguggi/oraclebeat/vendor/github.com/elastic/beats/metricbeat/mb/module/wrapper.go:106 +0x146
```

I've already opened the issue on the driver repository [#33](https://github.com/go-goracle/goracle/issues/33) but in the driver code `conn.go` the last call ist the the C function `dpiConn_release`. 

When I execute the procedure above and create an tcpdump on my interface and port 1521 I receive the following line:

```
153 31.614966 192.168.12.12 172.16.1.108 TCP 67 57578 → 1521 [PSH, ACK, URG] Seq=2790 Ack=87582 Win=182272 Urg=1 Len=1 TSval=32936936 TSecr=717866813[Malformed Packet: TNS]
```
The URG flag is set when I pressing `Ctrl+C` ...

But the curious thing is that this code works when the database is running on the same host. It looks like a vpn tunnel or firewall issue ... maybe???

Additionally the code works when I remove the `db.Close()` stuff from the ´Close()` method, but this is probably not the best idea. 

Is there a way or a procedure how to activate a trace so that I can further narrow the root cause?

**Comment by anthony-tuininga** [comment]
@cjbj may be able to help in this matter. It may be related to signal handler interference but I'll defer to Chris since he has a better understanding of these things!

**Comment by mguggi** [comment]
In the meantime we've found, that the most firewall providers don't accept the `URG` flag by default. Below some examples: 
 
1. [Check Point](https://supportcenter.checkpoint.com/supportcenter/portal?eventSubmit_doGoviewsolutiondetails=&solutionid=sk36869)
2. [Cisco](https://supportforums.cisco.com/t5/firewalling/asa-oracle-sqlnet-disconnects/td-p/2164919)

On our side we use `Check Point` and we can configure an exception on our firewall, but the likelihood that there is no firewall on the client side or only firewalls which accepting this flag is very low.

In my understanding, the `ODPI-C` library does not provide the `TCP` routines, but rather the `Oracle` driver libraries. If that's ture,  we can't do anything - right?

1. Why does `Oracle` use this flag when it comes to problems with most firewalls?
2. Is this behavior comprehensible or am I wrong with my findings?

**Comment by mguggi** [comment]
After my final analyze I come to the end that the `TCP` segments with urgent pointer was stripped by the firewalls. 

After some time of researching I found the Document [DISABLE_OOB (Out Of Band Break)? (Doc ID 373475.1)](https://support.oracle.com/epmos/faces/DocumentDisplay?_afrLoop=303065102840501&id=373475.1&_afrWindowMode=0&_adf.ctrl-state=1sccj07g1_4) on Oracle support site.

For all persons who have no access to the support site, a short excerpt from the document.  

> Data exception or break is a function in Oracle NET that allows a transaction to be interrupted before it is completed ... 
> 
> Out Of Band Breaks (OOB) are enabled by default provided the underlying protocol supports sending urgent data. 
>  
>  
> If the parameter DISABLE_OOB is set to OFF then it enables Oracle Net to send and receive "break" messages using urgent data provided by the underlying protocol. 
>  
>  
> If turned on, disables the ability to send and receive "break" messages using urgent data provided by the underlying protocol. 

In short, the solution is to set the `DISABLE_OOB=ON` parameter in the sqlnet.ora file.

**Comment by anthony-tuininga** [comment]
Thanks for the follow up. That will be helpful for anyone else who runs into the same issues!

**Comment by marksteward** [comment]
As reassurance, I've been using the same solution with a similarly awkward firewall for a few years and it doesn't seem to have any side-effects.
